server.c

```diff
@@ -0,0 +1,287 @@
+#include "server.h"
+
+#include "brute.h"
+#include "common.h"
+#include "multi.h"
+#include "thread_pool.h"
+
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+// TODO: Add status checks and cleanup in case of errors
+
+static status_t
+serv_context_init (serv_context_t *context, config_t *config)
+{
+  if (mt_context_init ((mt_context_t *)context, config) == S_FAILURE)
+    return (S_FAILURE);
+
+  if ((context->socket_fd = socket (AF_INET, SOCK_STREAM, 0)) == -1)
+    {
+      print_error ("Could not initialize server socket\n");
+      return (S_FAILURE);
+    }
+
+  int option = 1;
+  setsockopt (context->socket_fd, SOL_SOCKET, SO_REUSEADDR, &option,
+              sizeof (option));
+
+  struct sockaddr_in serv_addr;
+  serv_addr.sin_family = AF_INET;
+  serv_addr.sin_addr.s_addr = inet_addr (config->addr);
+  serv_addr.sin_port = htons (config->port);
+
+  if (bind (context->socket_fd, (struct sockaddr *)&serv_addr,
+            sizeof (serv_addr))
+      == -1)
+    {
+      print_error ("Could not bind socket\n");
```

Comment:
need to close (socket_fd)
It make sense to make a failure section at the end of the function